CNDB-10966 main: Add ability to read the TTLs and write times of the elements of a collection/UDT

adelapena opened the pull request:
Port of [DB-1289](https://datastax.jira.com/browse/DB-1289)/[CASSANDRA-8877](https://issues.apache.org/jira/browse/CASSANDRA-8877).

ekaterinadimitrova2 reviewed (approved):
Hey, great job! I spent quite some time reviewing this today as I looked into this patch for the first time.
I left a few comments to confirm my understanding. 
I checked in detail the few things that differentiated from the original code, like the lack of `InputRow' and things seemed okay. Test coverage looks solid, though Solar reports 87%. I don't know what our threshold is for this. On a quick look, it complained that we do not cover methods that throw `UnsupportedOperation` for instance. 

Even if we miss some corner cases somewhere, we cover everything in Apache and DSE, so there should be no regression.

I found this Butler diff. I am not sure it is the right link, but Butler did not link anything here. (Or did I not see it?)
http://butler-stargazer.aws.dsinternal.org/#/ci/upstream/compare/ds-cassandra-pr-gate/CNDB-10966-main/to/ds-cassandra-build-nightly/main

I checked the test marked as "regression," and I am fairly confident I've seen them before. Plus, they seem to be related to SAI and not to what we did here. 

Approving in case it satisfies our Solar metrics and CI is ok.

adelapena commented on src/java/org/apache/cassandra/cql3/selection/WritetimeOrTTLSelector.java:
```diff
@@ -19,120 +19,157 @@
 final class WritetimeOrTTLSelector extends Selector
 {
-    private final ColumnMetadata column;
-    private final int idx;
```
Right, it's unused

adelapena commented on src/java/org/apache/cassandra/cql3/selection/WritetimeOrTTLSelector.java:
```diff
@@ -19,120 +19,157 @@
+
+        WritetimeOrTTLSelector s = (WritetimeOrTTLSelector) o;
+
+        return Objects.equal(selected, s.selected) && kind == s.kind;
```
Good catch, added.

adelapena commented on src/java/org/apache/cassandra/cql3/selection/ColumnTimestamps.java:
```diff
@@ -0,0 +1,394 @@
+     */
+    private static class SingleTimestamps extends ColumnTimestamps
+    {
+        protected long timestamp;
```
Indeed.

adelapena commented on src/java/org/apache/cassandra/serializers/ListSerializer.java:
```diff
@@ -201,7 +206,22 @@ public ByteBuffer getSliceFromSerialized(ByteBuffer collection,
+    }
+
+    @Override
+    public int getIndexFromSerialized(ByteBuffer collection, ByteBuffer key, AbstractType<?> comparator)
```
Yes, we don't support that for lists. There is a mention about it in the super method.

adelapena commented on src/java/org/apache/cassandra/serializers/CollectionSerializer.java:
```diff
@@ -182,6 +184,37 @@ public abstract ByteBuffer getSliceFromSerialized(ByteBuffer collection,
+     * @param collection The serialized collection. This cannot be {@code null}.
+     * @param from  The left bound of the slice to extract. This cannot be {@code null} but if this is
+     * {@link ByteBufferUtil#UNSET_BYTE_BUFFER}, then the returned slice starts at the beginning of the collection.
+     * @param to The left bound of the slice to extract. This cannot be {@code null} but if this is
```
Eagle eye! Fixed.

ekaterinadimitrova2 replied:
I was trying to understand it and to understand the test results :D Those bounds are always confusing me :D

adelapena commented:
Thanks for the review. 

I think I have addressed them all te comments. 

I have also fixed some of the Sonar issues, but not the ones about the complexity of the `AbstractMapSerializer` methods. We have been using those for a long time and here we are just moving them around, so I think it's probably better not to touch them too much.

Regarding coverage, I don't know what's the threshold but it seems green at 87.6% in the report. Some of those methods are throwing assertion/unsupported errors because they are supposed to be unreachable. It doesn't seem very useful to create an artificial test to force reaching them.
